### PR TITLE
Update the TLS cipher string and provide configurability for TLS on outgoing federation

### PR DESCRIPTION
**Pull request by hawkowl**
As Twisted is updating the default to TLS 1.2 soon, Synapse needs a config option if any users wish to allow to connect to TLS 1.0/1.1 servers. This is probably not a problem in real world federation, but the option to set a higher minimum (1.3) in private federation setups is a valuable security knob.

### CODE REVIEW

**Comment by hawkowl** [comment]
Fixes https://github.com/matrix-org/synapse/issues/5529

File: synapse/config/tls.py [hunk 1]
```diff
@@ -81,6 +81,28 @@ def read_config(self, config, config_dir_path, **kwargs):
             "federation_verify_certificates", True
         )
 
+        # Minimum TLS version to use for outbound federation traffic
+        self.federation_minimum_tls_client_version = self.parse_float(
```
Comment (erikjohnston) [inline synapse/config/tls.py]: TBH I'd rather parse these as strings than worry about approximate float stuff. We don't seem to actually use the fact they're floats?

File: synapse/crypto/context_factory.py [hunk 2]
```diff
@@ -43,16 +56,18 @@ def configure_context(context, config):
...
+        context.set_options(
+            SSL.OP_NO_SSLv2 | SSL.OP_NO_SSLv3 | SSL.OP_NO_TLSv1 | SSL.OP_NO_TLSv1_1
+        )
         context.use_certificate_chain_file(config.tls_certificate_file)
         context.use_privatekey(config.tls_private_key)
 
         # https://hynek.me/articles/hardening-your-web-servers-ssl-ciphers/
         context.set_cipher_list(
-            "ECDH+AESGCM:ECDH+CHACHA20:ECDH+AES256:ECDH+AES128:!aNULL:!SHA1"
+            "ECDH+AESGCM:ECDH+CHACHA20:ECDH+AES256:ECDH+AES128:!aNULL:!SHA1:!AESCCM"
```
Comment (erikjohnston) [inline synapse/crypto/context_factory.py]: Is this a bad security choice?

File: synapse/crypto/context_factory.py [hunk 3]
```diff
@@ -43,16 +56,18 @@ def configure_context(context, config):
         try:
             _ecCurve = crypto.get_elliptic_curve(_defaultCurveName)
             context.set_tmp_ecdh(_ecCurve)
-
         except Exception:
             logger.exception("Failed to enable elliptic curve for TLS")
-        context.set_options(SSL.OP_NO_SSLv2 | SSL.OP_NO_SSLv3)
+
+        context.set_options(
+            SSL.OP_NO_SSLv2 | SSL.OP_NO_SSLv3 | SSL.OP_NO_TLSv1 | SSL.OP_NO_TLSv1_1
```
Comment (erikjohnston) [inline synapse/crypto/context_factory.py]: Why are we setting `NO_TLSv1`? I though the point was to optionally support it?

File: tests/config/test_tls.py [hunk 4]
```diff
@@ -78,3 +82,97 @@ def test_warn_self_signed(self):
...
+        """
+        config = {"federation_minimum_tls_client_version": 1.2}
+        t = TestConfig()
+        t.read_config(config, config_dir_path="", data_dir_path="")
+
+        cf = ClientTLSOptionsFactory(t)
+
+        # The context has had NO_TLSv1_1 set, but not NO_TLSv1_2
+        self.assertNotEqual(cf._verify_ssl._options & SSL.OP_NO_TLSv1_1, 0)
+        self.assertEqual(cf._verify_ssl._options & SSL.OP_NO_TLSv1_2, 0)
```
Comment (erikjohnston) [inline tests/config/test_tls.py]: Can we add a test that passing `1.0` works?

File: synapse/crypto/context_factory.py (hunk 2 above)
Comment (hawkowl) [reply synapse/crypto/context_factory.py]: AES128-CCM-8 is "not the best", AES128-CCM is just slow. They're not good ciphers that anything will ever pick, and advertising them makes our cipher string longer. (OpenSSL 1.1.1 started advertising it by default, so this update just makes it the same ciphers as when this originally landed)

File: synapse/crypto/context_factory.py (hunk 3 above)
Comment (hawkowl) [reply synapse/crypto/context_factory.py]: This is server, not client. Our server cipher string requires TLS 1.2 anyway to negotiate any of the ciphers, so this just makes the handshake a little simpler.

File: tests/config/test_tls.py (hunk 4 above)
Comment (hawkowl) [reply tests/config/test_tls.py]: done

File: synapse/config/tls.py (hunk 1 above)
Comment (hawkowl) [reply synapse/config/tls.py]: yeah switched to making it a string

**Comment by erikjohnston** [comment]
I think this is probably fine? Would be interested if @richvdh has any thoughts, as someone who has spent more time than me staring at TLS.

File: synapse/config/tls.py [hunk 5]
```diff
@@ -261,6 +282,15 @@ def generate_config_section(
         #
         #federation_verify_certificates: false
 
+        # The minimum TLS version that will be used for outbound federation requests.
+        #
+        # Defaults to `1`. Configurable to `1`, `1.1`, `1.2`, or `1.3`. Note
+        # that setting this value higher than `1.2` will prevent federation to most
+        # of the public Matrix network, only configure it to `1.3` if you have an
```
Comment (richvdh) [inline synapse/config/tls.py]:
```suggestion
        # of the public Matrix network: only configure it to `1.3` if you have an
```

File: synapse/config/tls.py [hunk 6]
```diff
@@ -261,6 +282,15 @@ def generate_config_section(
...
+        # entirely private federation setup and you can ensure TLS 1.3 support.
+        #
+        #federation_minimum_tls_client_version: 1.2
```
Comment (richvdh) [inline synapse/config/tls.py]: I think I'd go with `federation_client_minimum_tls_version`, but w/e

**Pull request merged**